### PR TITLE
Add several new layers to be able to support Resnet like architectures

### PR DESCRIPTION
**Pull request by franckma31**
* Add support and unit tests for replicate padding in convolution (unconstrained.py)
* Add support and unit tests for LipResidual (residual.py)
* Modify InvertibleDownSampling and InvertibleUpSampling to use Pixel(Un)Shuffle pytorch functions
* Add support and unit tests for SpectralConv1d (conv.py)
*  Add support and unit tests  for HouseHolder activation (activation.py)
* Add support and unit tests for SpectralConv1d (conv.py)
* Add support and unit tests for BatchCentering and LayerCentering (normalization.py)
* Add support and unit tests for SpectralConvTranspose2d  (conv.py)
* Add support and unit tests for ScaledGlobalL2NormPool2d (pooling.py)
* Update update groupsort and groupsort2 activation to be consistent with deellip (actication.py)

* Add a Warning in notebooks to support vanilla_export after loading weights, to  solve the issue #17

### CODE REVIEW

File: deel/torchlip/modules/activation.py [hunk 1]
```diff
@@ -211,3 +212,55 @@ def vanilla_export(self):
...
+                ), f"Unknown initializer {theta_initializer}"
+                name2init[theta_initializer](self.theta)
+            elif isinstance(theta_initializer, float):
+                pass
+            else:
+                raise ValueError(f"Unknown initializer {theta_initializer}")
+
+    def forward(self, z, axis=1):
+        theta_shape = (1, -1) + (1,) * (len(z.shape) - 2)
+        theta = self.theta.to(z.device).view(theta_shape)
```
Comment (thib-s) [inline deel/torchlip/modules/activation.py:254]: I think we can remove the transfert `.to(z.device)` since the module parameter should be already transfered on GPU/cpu when model is declared on GPU/cpu

File: deel/torchlip/modules/pooling.py [hunk 2]
```diff
@@ -204,3 +204,56 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
 
     def vanilla_export(self):
         return self
+
+
+class ScaledGlobalL2NormPool2d(torch.nn.AdaptiveAvgPool2d, LipschitzModule):
```
Comment (thib-s) [inline deel/torchlip/modules/pooling.py:209]: torch LPPool2D internally uses `torch.nn.functional.lp_pool2d`, shouldn't we use the same function in this layer?

File: deel/torchlip/utils/lconv_norm.py [hunk 3]
```diff
@@ -24,24 +24,54 @@
...
+    kernel_size: Tuple[int],
+    input_shape: Tuple[int] = None,
+    strides: Tuple[int] = (1,),
+    padding_mode: str = "zeros",
+) -> float:
+    stride = strides[0]
+    k1 = kernel_size[0]
+
+    if (
+        (padding_mode in ["zeros", "same"])
```
Comment (thib-s) [inline deel/torchlip/utils/lconv_norm.py:45]: shouldn't be `["zeros", "circular"]` ? 
```suggestion
        (padding_mode in ["zeros", "circular"])
```

File: deel/torchlip/utils/lconv_norm.py [hunk 4]
```diff
@@ -24,24 +24,54 @@
...
+        and (stride == 1)
+        and (input_shape is not None)
+    ):
+        # See https://arxiv.org/abs/2006.06520
+        in_l = input_shape[-1]
+        k1_div2 = (k1 - 1) / 2
+        coefLip = in_l / (k1 * in_l - k1_div2 * (k1_div2 + 1))
+    else:
```
Comment (thib-s) [inline deel/torchlip/utils/lconv_norm.py:53]: this condition works for circular padding but probably not for "symmetric" or "reflect".

File: deel/torchlip/utils/lconv_norm.py [hunk 5]
```diff
@@ -24,24 +24,54 @@
...
     strides: Tuple[int, ...] = (1, 1),
+    padding_mode: str = "zeros",
 ) -> float:
     # See https://arxiv.org/abs/2006.06520
     stride = np.prod(strides)
     k1, k2 = kernel_size
 
-    if stride == 1 and input_shape is not None:
+    if (
+        (padding_mode in ["zeros", "same"])
```
Comment (thib-s) [inline deel/torchlip/utils/lconv_norm.py:71]: see previous comment

File: deel/torchlip/modules/activation.py [hunk 6]
```diff
@@ -211,3 +212,55 @@ def vanilla_export(self):
         layer = LPReLU(num_parameters=self.num_parameters)
         layer.weight.data = self.weight.data
         return layer
+
+
+class HouseHolder(nn.Module, LipschitzModule):
```
Comment (thib-s) [inline deel/torchlip/modules/activation.py:217]: should we add HouseHolderOrder2 also ?

**Review by thib-s — COMMENTED** [review]
See my comments
Also, tests seem to break for some versions.

File: deel/torchlip/modules/normalization.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        if bias:
+            self.bias = nn.Parameter(torch.zeros((size,)), requires_grad=True)
+        else:
+            self.register_parameter("bias", None)
+        self.dim = dim
+
+    def forward(self, x):
+        mean = x.mean(dim=self.dim, keepdim=True)
+        if self.bias is not None:
+            bias_shape = (1, -1) + (1,) * (len(x.shape) - 2)
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/torchlip/modules/normalization.py:46]: I don't think so, we have one bias per channel (size parameter) and thus we need to have len(x.shape) dimensions, but two are given by (1,-1). This is independent to the self.dim parameter for the mean  (even if by default the average is done on all but the two first layers also).

File: deel/torchlip/utils/lconv_norm.py (hunk 3 above)
Comment (franckma31) [reply deel/torchlip/utils/lconv_norm.py]: accepted in f3f32bbf560b48d3e96af79e931d0951afc710db

File: deel/torchlip/utils/lconv_norm.py (hunk 5 above)
Comment (franckma31) [reply deel/torchlip/utils/lconv_norm.py]: accepted in f3f32bbf560b48d3e96af79e931d0951afc710db

File: deel/torchlip/modules/activation.py (hunk 1 above)
Comment (franckma31) [reply deel/torchlip/modules/activation.py]: accepted in 34105b9cb838ea3a3d7e0346322198b77bdbbe3b

File: deel/torchlip/modules/activation.py (hunk 6 above)
Comment (franckma31) [reply deel/torchlip/modules/activation.py:217]: Could be if useful. But in a new PR?

File: deel/torchlip/utils/lconv_norm.py (hunk 4 above)
Comment (franckma31) [reply deel/torchlip/utils/lconv_norm.py:49]: Condition for applying the factor proposed in are the following:
(padding_mode in ["zeros"]) and (stride == 1) and (input_shape is not None)
Otherwise the factor is k_size/stride

As discussed, symmetric padding with RKO is k_size Lipschitz.
But reflect padding could be more thant k_size. We may reject this padding in SpectralConv layer ?

File: deel/torchlip/modules/normalization.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (thib-s) [reply deel/torchlip/modules/normalization.py:46]: I agree

File: deel/torchlip/utils/lconv_norm.py (hunk 4 above)
Comment (thib-s) [reply deel/torchlip/utils/lconv_norm.py:49]: Agreed

**Review by thib-s — APPROVED** [review]
LGTM
